File: Javascript/Leetcode/Eventual Safe States.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function (graph) {
  const n = graph.length;
  let adj = Array.from({ length: n }, () => []);

  let ingoing = Array(n).fill(0);

  for (let i = 0; i < graph.length; i++) {
    for (let j = 0; j < graph[i].length; j++) {
      adj[graph[i][j]].push(i);
      ingoing[i]++;
    }
  }

  let p = new Set();

  for (let i = 0; i < ingoing.length; i++) {
    if (!ingoing[i]) {
      p.add(i);
    }
  }
  let ans = [];
  while (!p.size) {
    let node = p.pop();
    ans.push(node);

    for (const nei of adj[node]) {
      ingoing[nei]--;
      if (!ingoing[i]) {
        p.add(nei);
      }
    }
  }
  ans.sort((a, b) => a - b);
  return ans;
};

// eventualSafeNodes();
let p = new Set([1, 2, 31, 4]);
let iterator = p.values(); // Get an iterator for the values of the Set

console.log(iterator.next()); // Output: { value: 1, done: false }
console.log(iterator.next()); // Output: { value: 2, done: false }
console.log(iterator.next()); // Output: { value: 3, done: false }
console.log(iterator.next()); // Output: { value: 3, done: false }
console.log(p.values()); 
console.log(iterator.next()); // Output: { value: undefined, done: true }

File: Leetcode/Questions/Graph/Problems On Topo Sort/Eventual_Safe_States.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> eventualSafeNodes(vector<vector<int>> &graph)
    {
        int n = graph.size();
        vector<int> adj[n];
        vector<int> ingoing(n, 0);
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < graph[i].size(); j++)
            {
                adj[graph[i][j]].push_back(i);
                ingoing[i]++;
            }
        }
        queue<int> pq;
        for (int i = 0; i < n; i++)
        {
            if (!ingoing[i])
            {
                pq.push(i);
            }
        }

        vector<int> ans;
        while (!pq.empty())
        {
            int node = pq.front();
            ans.push_back(node);
            pq.pop();

            for (auto &&nei : adj[node])
            {
                ingoing[nei]--;
                if (!ingoing[nei])
                {
                    pq.push(nei);
                }
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

int main()
{

    return 0;
}